File: npm/src/rules/quality.ts
import type { PromptlintConfig } from "../config";
import type { Finding } from "./cost";
import { preview, lineNumber, lineContext } from "./utils";

export function checkStructure(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["structure-sections"] ?? true)) return [];

  const hasXmlTags = /<\w+>/.test(text);
  const hasHeadings = /(?:^|\n)(?:Task|Context|Output|Goal|Requirements|Instructions):\s/im.test(text);
  const hasMarkdownHeaders = /(?:^|\n)#{1,6}\s+\w+/.test(text);
  const hasJsonStructure =
    (text.includes("{") || text.includes("[")) &&
    (text.includes("}") || text.includes("]")) &&
    (text.includes('"') || text.includes("'"));
  const hasDelimiters = /```|^---\s*$/m.test(text);
  const hasNumberedSections = /(?:^|\n)\d+\.\s+\w+/.test(text);

  if (hasXmlTags || hasHeadings || hasMarkdownHeaders || hasJsonStructure || hasDelimiters || hasNumberedSections) {
    return [];
  }

  return [
    {
      level: "WARN",
      rule: "structure-sections",
      message: "No explicit sections detected (Task/Context/Output).",
      line: "-",
      context: preview(text, config.previewLength),
    },
    {
      level: "INFO",
      rule: "structure-recommendations",
      message: "Recommended templates: headings (Task:, Context:, Output:) / XML tags (<task>) / Markdown (## sections).",
      line: "-",
      context: preview(text, config.previewLength),
    },
  ];
}

// Bug 1.3: Tightened vague-term patterns — removed good/bad/nice/better/best,
// appropriate/suitable/relevant/proper which caused high false-positive rates.
export function checkClarity(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["clarity-vague-terms"] ?? true)) return [];

  const vaguePatterns: [RegExp, string][] = [
    [/\b(some|several|various|many|few)\b/gi, "vague quantifier — specify a number or range"],
    [/\b(stuff|things|etc|and so on|and more)\b/gi, "trailing vague catch-all — enumerate explicitly"],
    [/\b(somehow|somewhere|sometime)\b/gi, "unspecified manner/place/time"],
    [/\bmaybe\b(?!\s+(?:null|undefined|none|empty))/gi, "uncertain language in an instruction context"],
  ];

  const results: Finding[] = [];
  for (const [re, issueType] of vaguePatterns) {
    let m: RegExpExecArray | null;
    re.lastIndex = 0;
    while ((m = re.exec(text)) !== null) {
      results.push({
        level: "WARN",
        rule: "clarity-vague-terms",
        message: `Vague term '${m[0]}' detected (${issueType}). Be more specific.`,
        line: lineNumber(text, m.index),
        context: lineContext(text, m.index, config.contextWidth),
        savings: 2.0,
      });
    }
  }
  return results;
}

export function checkSpecificity(text: string, config: PromptlintConfig): Finding[] {
  const results: Finding[] = [];

  if (config.enabledRules["specificity-examples"] ?? true) {
    const hasInstruction = /\b(write|create|generate|build|implement|design)\b/i.test(text);
    const hasExample = /\b(example|e\.g\.|such as|like|for instance)\b/i.test(text);
    if (hasInstruction && !hasExample && text.length > 100) {
      results.push({
        level: "INFO",
        rule: "specificity-examples",
        message: "Consider adding examples to clarify expected output format.",
        line: "-",
        context: preview(text, config.previewLength),
      });
    }
  }

  if (config.enabledRules["specificity-constraints"] ?? true) {
    const hasTask = /\b(write|create|generate|list|explain)\b/i.test(text);
    const hasConstraint = /\b(must|should|limit|maximum|minimum|between|exactly|only)\b/i.test(text);
    if (hasTask && !hasConstraint && text.length > 80) {
      results.push({
        level: "INFO",
        rule: "specificity-constraints",
        message: "Consider adding constraints (length, format, scope) for clearer results.",
        line: "-",
        context: preview(text, config.previewLength),
      });
    }
  }

  return results;
}

// Bug 1.4: Smarter sentence splitting avoiding abbreviations (e.g., i.e., decimals).
// Bug 1.7: Expanded redundancy patterns (7 → 41).
export function checkVerbosity(text: string, config: PromptlintConfig): Finding[] {
  const results: Finding[] = [];

  if (config.enabledRules["verbosity-sentence-length"] ?? true) {
    // Split on sentence-ending punctuation followed by whitespace+capital,
    // avoiding splits on abbreviations (e.g., v1.0, decimals, Mr. etc.)
    const sentences = text.split(/(?<=[^A-Z\d][.!?])\s+(?=[A-Z])|(?<=[.!?]{2})\s+/);
    for (const sentence of sentences) {
      const words = sentence.trim().split(/\s+/).filter(Boolean);
      if (words.length > 40) {
        results.push({
          level: "INFO",
          rule: "verbosity-sentence-length",
          message: `Long sentence detected (${words.length} words). Consider breaking it up.`,
          line: "-",
          context: preview(sentence.trim(), config.previewLength),
          savings: 3.0,
        });
      }
    }
  }

  if (config.enabledRules["verbosity-redundancy"] ?? true) {
    const redundantPatterns: RegExp[] = [
      /\bin order to\b/gi,
      /\bdue to the fact that\b/gi,
      /\bat this point in time\b/gi,
      /\bfor the purpose of\b/gi,
      /\bin the event that\b/gi,
      /\bprior to\b/gi,
      /\bsubsequent to\b/gi,
      /\ba total of\b/gi,
      /\beach and every\b/gi,
      /\bfirst and foremost\b/gi,
      /\bfuture plans\b/gi,
      /\bpast history\b/gi,
      /\bend result\b/gi,
      /\bbasic fundamentals\b/gi,
      /\bclose proximity\b/gi,
      /\bgather together\b/gi,
      /\bjoin together\b/gi,
      /\brefer back\b/gi,
      /\breturn back\b/gi,
      /\bunexpected surprise\b/gi,
      /\bcompletely eliminate\b/gi,
      /\bcompletely finished\b/gi,
      /\badvance planning\b/gi,
      /\bpast experience\b/gi,
      /\bnew innovation\b/gi,
      /\bpersonal opinion\b/gi,
      /\brepeat again\b/gi,
      /\bstill remains\b/gi,
      /\btrue fact\b/gi,
      /\bwith the exception of\b/gi,
      /\bin close proximity to\b/gi,
      /\bhas the ability to\b/gi,
      /\bis able to\b/gi,
      /\bin spite of the fact that\b/gi,
      /\bwith regard to\b/gi,
      /\bin relation to\b/gi,
      /\bfor the reason that\b/gi,
      /\bin the near future\b/gi,
      /\bat the present time\b/gi,
      /\buntil such time as\b/gi,
      /\bon a (?:daily|weekly|monthly) basis\b/gi,
    ];
    for (const re of redundantPatterns) {
      re.lastIndex = 0;
      const m = re.exec(text);
      if (m) {
        results.push({
          level: "INFO",
          rule: "verbosity-redundancy",
          message: `Redundant phrase '${m[0]}' detected. Use simpler alternative.`,
          line: lineNumber(text, m.index),
          context: lineContext(text, m.index, config.contextWidth),
          savings: 2.0,
        });
      }
    }
  }

  return results;
}

// Bug 1.5: Add real weak-verb detection; raise passive threshold from 3 to 5.
const WEAK_VERB_PATTERNS: [RegExp, string][] = [
  [/\b(consider|try to|attempt to|endeavor to)\b/gi, "weak directive — use imperative form"],
  [/\byou might want to\b/gi, "hedged instruction — use direct imperative"],
  [/\bfeel free to\b/gi, "unnecessary hedge — remove"],
  [/\bit would be (?:good|nice|helpful|great) (?:if|to)\b/gi, "indirect instruction — state directly"],
  [/\bif possible\b/gi, "conditional hedge — commit to the instruction"],
  [/\bwhenever (?:possible|you can)\b/gi, "weak conditional — use 'always' or state the condition explicitly"],
];

export function checkActionability(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["actionability-weak-verbs"] ?? true)) return [];

  const results: Finding[] = [];

  for (const [re, issueType] of WEAK_VERB_PATTERNS) {
    re.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      results.push({
        level: "INFO",
        rule: "actionability-weak-verbs",
        message: `Weak directive '${m[0]}' (${issueType}).`,
        line: lineNumber(text, m.index),
        context: lineContext(text, m.index, config.contextWidth),
      });
    }
  }

  const passiveRe = /\b(is|are|was|were|be|been|being)\s+\w+ed\b/gi;
  const passiveMatches = [...text.matchAll(passiveRe)];
  if (passiveMatches.length > 5) {
    results.push({
      level: "INFO",
      rule: "actionability-weak-verbs",
      message: `Multiple passive voice constructions (${passiveMatches.length}) detected. Use active voice for clarity.`,
      line: "-",
      context: preview(text, config.previewLength),
    });
  }
  return results;
}

// Bug 1.6: Expanded from 3 to 12 built-in term pairs; custom_term_pairs support.
const BUILTIN_TERM_PAIRS: [RegExp, RegExp][] = [
  [/\buser\b/i, /\bcustomer\b/i],
  [/\bfunction\b/i, /\bmethod\b/i],
  [/\berror\b/i, /\bexception\b/i],
  [/\bAI\b/, /\bmodel\b/i],
  [/\bLLM\b/, /\bmodel\b/i],
  [/\bquery\b/i, /\brequest\b/i],
  [/\bresponse\b/i, /\breply\b/i],
  [/\boutput\b/i, /\bresult\b/i],
  [/\bprompt\b/i, /\bmessage\b/i],
  [/\bsystem prompt\b/i, /\binstruction\b/i],
  [/\btask\b/i, /\bgoal\b/i],
  [/\bagent\b/i, /\bassistant\b/i],
];

export function checkConsistency(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["consistency-terminology"] ?? true)) return [];

  const results: Finding[] = [];

  function checkPair(re1: RegExp, re2: RegExp): void {
    const m1 = re1.exec(text);
    const m2 = re2.exec(text);
    if (m1 && m2) {
      results.push({
        level: "INFO",
        rule: "consistency-terminology",
        message: `Mixed terminology: '${m1[0]}' and '${m2[0]}'. Use one term consistently.`,
        line: "-",
        context: preview(text, config.previewLength),
      });
    }
  }

  for (const [re1, re2] of BUILTIN_TERM_PAIRS) {
    checkPair(re1, re2);
  }

  for (const group of (config.customTermPairs ?? [])) {
    for (let i = 0; i < group.length; i++) {
      for (let j = i + 1; j < group.length; j++) {
        const esc = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        checkPair(new RegExp(`\\b${esc(group[i])}\\b`, "i"), new RegExp(`\\b${esc(group[j])}\\b`, "i"));
      }
    }
  }

  return results;
}

export function checkCompleteness(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["completeness-edge-cases"] ?? true)) return [];

  const hasTask = /\b(write|create|implement|build|generate)\b/i.test(text);
  const hasEdgeCases = /\b(edge case|error|exception|invalid|empty|null|missing)\b/i.test(text);
  if (hasTask && !hasEdgeCases && text.length > 100) {
    return [{
      level: "INFO",
      rule: "completeness-edge-cases",
      message: "Consider specifying how to handle edge cases and errors.",
      line: "-",
      context: preview(text, config.previewLength),
    }];
  }
  return [];
}

// ── Rule 2.2: Role clarity ─────────────────────────────────────────────────

export function checkRoleClarity(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["role-clarity"] ?? true)) return [];
  if (text.split(/\s+/).length < 30) return [];

  const isInstructional = /\b(you|your|respond|answer|help|assist|always|never|must|should)\b/i.test(text);
  if (!isInstructional) return [];

  const hasRole = /\b(you are|you're|act as|your role is|you will serve as|you are a|you are an|you work as|your job is|your task is|you are responsible for|you specialize in)\b/i.test(text);
  if (!hasRole) {
    return [{
      level: "WARN",
      rule: "role-clarity",
      message: "No role or persona defined. Add 'You are a [role]...' to improve output consistency.",
      line: "-",
      context: preview(text, config.previewLength),
    }];
  }
  return [];
}

// ── Rule 2.3: Output format missing ───────────────────────────────────────

const OUTPUT_INSTRUCTION_VERBS = /\b(list|extract|find|return|give me|provide|output|generate|create|show|summarize|identify|enumerate)\b/i;
const OUTPUT_FORMAT_KEYWORDS = /\b(JSON|XML|CSV|markdown|bullet|numbered|table|plain text|HTML|YAML|format:|return as|output as|structured as|schema)\b/i;

export function checkOutputFormat(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["output-format-missing"] ?? true)) return [];
  if (text.length < 60) return [];

  if (OUTPUT_INSTRUCTION_VERBS.test(text) && !OUTPUT_FORMAT_KEYWORDS.test(text)) {
    return [{
      level: "WARN",
      rule: "output-format-missing",
      message: "Output instruction detected but no format specified (JSON, markdown, bullet list, etc.). Unspecified format produces inconsistent results.",
      line: "-",
      context: preview(text, config.previewLength),
    }];
  }
  return [];
}

// ── Rule 2.6: Hallucination risk ──────────────────────────────────────────

const FACTUAL_PATTERNS = [
  /\b(what is|what are|who is|who are|when did|when was|where is|how many|how much)\b/i,
  /\b(current(ly)?|latest|recent(ly)?|today|now|as of|up to date)\b/i,
  /\b(tell me about|give me (?:the|a) (?:list|summary|overview) of)\b/i,
];

const GROUNDING_PATTERNS = [
  /\{[\w\s]+\}/,
  /<context>/i,
  /Context:/i,
  /```/,
  /Given the following/i,
  /Based on the (?:following|above|provided)/i,
  /Using the (?:data|information|context|text) (?:below|above|provided)/i,
];

export function checkHallucinationRisk(text: string, config: PromptlintConfig): Finding[] {
  if (!(config.enabledRules["hallucination-risk"] ?? true)) return [];

  const hasFactual = FACTUAL_PATTERNS.some((re) => re.test(text));
  const hasGrounding = GROUNDING_PATTERNS.some((re) => re.test(text));

  if (hasFactual && !hasGrounding) {
    return [{
      level: "WARN",
      rule: "hallucination-risk",
      message: "Prompt requests factual/current information without grounding context. Consider adding a {context} variable or <context> section with source data.",
      line: "-",
      context: preview(text, config.previewLength),
    }];
  }
  return [];
}

File: cli/promptlint/rules/quality.py
from __future__ import annotations

import re
from typing import Dict, List

from ..utils.config import PromptlintConfig


def _preview(text: str, limit: int) -> str:
    cleaned = " ".join(text.split())
    return cleaned[:limit] + ("..." if len(cleaned) > limit else "")


def _line_number(text: str, index: int) -> int:
    return text.count("\n", 0, index) + 1


def _line_context(text: str, index: int, width: int) -> str:
    line_start = text.rfind("\n", 0, index) + 1
    line_end = text.find("\n", index)
    if line_end == -1:
        line_end = len(text)
    line = text[line_start:line_end]
    column = index - line_start

    if len(line) > width:
        half = width // 2
        left = max(column - half, 0)
        right = min(left + width, len(line))
        if right - left < width:
            left = max(right - width, 0)
        trimmed = line[left:right]
        caret_pos = column - left
        prefix = "..." if left > 0 else ""
        suffix = "..." if right < len(line) else ""
        display_line = f"{prefix}{trimmed}{suffix}"
        caret_pos += len(prefix)
    else:
        display_line = line
        caret_pos = column

    return f"{display_line}\n{' ' * caret_pos}^"


def check_structure(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if not config.enabled_rules.get("structure-sections", True):
        return results

    has_xml_tags = bool(re.search(r"<\w+>", text))
    has_headings = bool(re.search(
        r"(?:^|\n)(?:Task|Context|Output|Goal|Requirements|Instructions):\s",
        text, re.IGNORECASE,
    ))
    has_markdown_headers = bool(re.search(r"(?:^|\n)#{1,6}\s+\w+", text))
    has_json_structure = ("{" in text or "[" in text) and ("}" in text or "]" in text) and ('"' in text or "'" in text)
    has_delimiters = bool(re.search(r"```|^---\s*$", text, re.MULTILINE))
    has_numbered_sections = bool(re.search(r"(?:^|\n)\d+\.\s+\w+", text))

    has_any_structure = (
        has_xml_tags
        or has_headings
        or has_markdown_headers
        or has_json_structure
        or has_delimiters
        or has_numbered_sections
    )

    if not has_any_structure:
        results.append(
            {
                "level": "WARN",
                "rule": "structure-sections",
                "message": "No explicit sections detected (Task/Context/Output).",
                "line": "-",
                "context": _preview(text, config.preview_length),
            }
        )
        results.append(
            {
                "level": "INFO",
                "rule": "structure-recommendations",
                "message": "Recommended templates: headings (Task:, Context:, Output:) / XML tags (<task>) / Markdown (## sections).",
                "line": "-",
                "context": _preview(text, config.preview_length),
            }
        )

    return results


def check_clarity(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if not config.enabled_rules.get("clarity-vague-terms", True):
        return results

    vague_patterns = [
        (r"\b(some|several|various|many|few)\b", "vague quantifier — specify a number or range"),
        (r"\b(stuff|things|etc|and so on|and more)\b", "trailing vague catch-all — enumerate explicitly"),
        (r"\b(somehow|somewhere|sometime)\b", "unspecified manner/place/time"),
        (r"\bmaybe\b(?!\s+(?:null|undefined|none|empty))", "uncertain language in an instruction context"),
    ]

    for pattern, issue_type in vague_patterns:
        for match in re.finditer(pattern, text, re.IGNORECASE):
            results.append(
                {
                    "level": "WARN",
                    "rule": "clarity-vague-terms",
                    "message": f"Vague term '{match.group(0)}' detected ({issue_type}). Be more specific.",
                    "line": _line_number(text, match.start()),
                    "context": _line_context(text, match.start(), config.context_width),
                    "savings": 2.0,
                }
            )

    return results


def check_specificity(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if config.enabled_rules.get("specificity-examples", True):
        has_instruction = bool(re.search(r"\b(write|create|generate|build|implement|design)\b", text, re.IGNORECASE))
        has_example = bool(re.search(r"\b(example|e\.g\.|such as|like|for instance)\b", text, re.IGNORECASE))

        if has_instruction and not has_example and len(text) > 100:
            results.append(
                {
                    "level": "INFO",
                    "rule": "specificity-examples",
                    "message": "Consider adding examples to clarify expected output format.",
                    "line": "-",
                    "context": _preview(text, config.preview_length),
                }
            )

    if config.enabled_rules.get("specificity-constraints", True):
        has_task = bool(re.search(r"\b(write|create|generate|list|explain)\b", text, re.IGNORECASE))
        has_constraint = bool(re.search(r"\b(must|should|limit|maximum|minimum|between|exactly|only)\b", text, re.IGNORECASE))

        if has_task and not has_constraint and len(text) > 80:
            results.append(
                {
                    "level": "INFO",
                    "rule": "specificity-constraints",
                    "message": "Consider adding constraints (length, format, scope) for clearer results.",
                    "line": "-",
                    "context": _preview(text, config.preview_length),
                }
            )

    return results


def check_verbosity(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if config.enabled_rules.get("verbosity-sentence-length", True):
        # Split on sentence-ending punctuation followed by whitespace + capital letter,
        # avoiding splits on abbreviations (e.g., i.e., v1.0, decimals).
        sentences = re.split(r'(?<=[^A-Z\d][.!?])\s+(?=[A-Z])|(?<=[.!?]{2})\s+', text)
        for sentence in sentences:
            words = sentence.split()
            if len(words) > 40:
                results.append(
                    {
                        "level": "INFO",
                        "rule": "verbosity-sentence-length",
                        "message": f"Long sentence detected ({len(words)} words). Consider breaking it up.",
                        "line": "-",
                        "context": _preview(sentence.strip(), config.preview_length),
                        "savings": 3.0,
                    }
                )

    if config.enabled_rules.get("verbosity-redundancy", True):
        redundant_patterns = [
            r"\bin order to\b",
            r"\bdue to the fact that\b",
            r"\bat this point in time\b",
            r"\bfor the purpose of\b",
            r"\bin the event that\b",
            r"\bprior to\b",
            r"\bsubsequent to\b",
            r"\ba total of\b",
            r"\beach and every\b",
            r"\bfirst and foremost\b",
            r"\bfuture plans\b",
            r"\bpast history\b",
            r"\bend result\b",
            r"\bbasic fundamentals\b",
            r"\bclose proximity\b",
            r"\bgather together\b",
            r"\bjoin together\b",
            r"\brefer back\b",
            r"\breturn back\b",
            r"\bunexpected surprise\b",
            r"\bcompletely eliminate\b",
            r"\bcompletely finished\b",
            r"\badvance planning\b",
            r"\bpast experience\b",
            r"\bnew innovation\b",
            r"\bpersonal opinion\b",
            r"\brepeat again\b",
            r"\bstill remains\b",
            r"\btrue fact\b",
            r"\bwith the exception of\b",
            r"\bin close proximity to\b",
            r"\bhas the ability to\b",
            r"\bis able to\b",
            r"\bin spite of the fact that\b",
            r"\bwith regard to\b",
            r"\bin relation to\b",
            r"\bfor the reason that\b",
            r"\bin the near future\b",
            r"\bat the present time\b",
            r"\buntil such time as\b",
            r"\bon a (?:daily|weekly|monthly) basis\b",
        ]

        for pattern in redundant_patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                results.append(
                    {
                        "level": "INFO",
                        "rule": "verbosity-redundancy",
                        "message": f"Redundant phrase '{match.group(0)}' detected. Use simpler alternative.",
                        "line": _line_number(text, match.start()),
                        "context": _line_context(text, match.start(), config.context_width),
                        "savings": 2.0,
                    }
                )

    return results


_WEAK_VERB_PATTERNS = [
    (r"\b(consider|try to|attempt to|endeavor to)\b", "weak directive — use imperative form"),
    (r"\byou might want to\b", "hedged instruction — use direct imperative"),
    (r"\bfeel free to\b", "unnecessary hedge — remove"),
    (r"\bit would be (?:good|nice|helpful|great) (?:if|to)\b", "indirect instruction — state directly"),
    (r"\bif possible\b", "conditional hedge — commit to the instruction"),
    (r"\bwhenever (?:possible|you can)\b", "weak conditional — use 'always' or state the condition explicitly"),
]


def check_actionability(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if not config.enabled_rules.get("actionability-weak-verbs", True):
        return results

    for pattern, issue_type in _WEAK_VERB_PATTERNS:
        for match in re.finditer(pattern, text, re.IGNORECASE):
            results.append(
                {
                    "level": "INFO",
                    "rule": "actionability-weak-verbs",
                    "message": f"Weak directive '{match.group(0)}' ({issue_type}).",
                    "line": _line_number(text, match.start()),
                    "context": _line_context(text, match.start(), config.context_width),
                }
            )

    passive_pattern = r"\b(is|are|was|were|be|been|being)\s+\w+ed\b"
    passive_matches = list(re.finditer(passive_pattern, text, re.IGNORECASE))

    if len(passive_matches) > 5:
        results.append(
            {
                "level": "INFO",
                "rule": "actionability-weak-verbs",
                "message": f"Multiple passive voice constructions ({len(passive_matches)}) detected. Use active voice for clarity.",
                "line": "-",
                "context": _preview(text, config.preview_length),
            }
        )

    return results


_BUILTIN_TERM_PAIRS = [
    (r"\buser\b", r"\bcustomer\b"),
    (r"\bfunction\b", r"\bmethod\b"),
    (r"\berror\b", r"\bexception\b"),
    (r"\bAI\b", r"\bmodel\b"),
    (r"\bLLM\b", r"\bmodel\b"),
    (r"\bquery\b", r"\brequest\b"),
    (r"\bresponse\b", r"\breply\b"),
    (r"\boutput\b", r"\bresult\b"),
    (r"\bprompt\b", r"\bmessage\b"),
    (r"\bsystem prompt\b", r"\binstruction\b"),
    (r"\btask\b", r"\bgoal\b"),
    (r"\bagent\b", r"\bassistant\b"),
]


def check_consistency(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if not config.enabled_rules.get("consistency-terminology", True):
        return results

    def _check_pair(p1: str, p2: str) -> None:
        m1 = re.search(p1, text, re.IGNORECASE)
        m2 = re.search(p2, text, re.IGNORECASE)
        if m1 and m2:
            results.append(
                {
                    "level": "INFO",
                    "rule": "consistency-terminology",
                    "message": f"Mixed terminology: '{m1.group(0)}' and '{m2.group(0)}'. Use one term consistently.",
                    "line": "-",
                    "context": _preview(text, config.preview_length),
                }
            )

    for p1, p2 in _BUILTIN_TERM_PAIRS:
        _check_pair(p1, p2)

    for group in config.custom_term_pairs:
        for i in range(len(group)):
            for j in range(i + 1, len(group)):
                _check_pair(rf"\b{re.escape(group[i])}\b", rf"\b{re.escape(group[j])}\b")

    return results


def check_completeness(text: str, config: PromptlintConfig) -> List[Dict[str, object]]:
    results: List[Dict[str, object]] = []

    if not config.enabled_rules.get("completeness-edge-cases", True):
        return results

    has_task = bool(re.search(r"\b(write|create|implement|build|generate)\b", text, re.IGNORECASE))
    has_edge_cases = bool(re.search(r"\b(edge case|error|exception|invalid|empty|null|missing)\b", text, re.IGNORECASE))

    if has_task and not has_edge_cases and len(text) > 100:
        results.append(
            {
                "level": "INFO",
                "rule": "completeness-edge-cases",
                "message": "Consider specifying how to handle edge cases and errors.",
                "line": "-",
                "context": _preview(text, config.preview_length),
            }
        )

    return results
